cpp/dolfinx/fem/Form.h

```diff
@@ -378,173 +474,77 @@ class Form
-  /// for a given ID in mixed-topology meshes).
-  /// @return List of active entities for the given integral (kernel).
-  std::vector<std::int32_t> domain(IntegralType type, int i,
-                                   int kernel_idx) const
+  /// @param[in] id Integral ID, i.e. (sub)domain index.
+  /// @param[in] kernel_idx Index of the kernel (we may have multiple
+  /// kernels for a given ID in mixed-topology meshes).
+  /// @return Entity indices in the 'integration domain' mesh.
+  std::span<const std::int32_t> domain(IntegralType type, int id,
+                                       int kernel_idx) const
   {
-    const std::vector<integral_data<scalar_type, geometry_type>>& integrals
-        = _integrals[static_cast<std::size_t>(type)];
-    auto get_id = [](const auto& a) { return a.id; };
-    auto start = std::ranges::lower_bound(integrals, i, std::less<>{}, get_id);
-    auto end = std::ranges::upper_bound(integrals, i, std::less<>{}, get_id);
-
-    // Check that the kernel is valid and return it if so
-    if (start == integrals.end() or start->id != i
-        or std::distance(start, end) <= kernel_idx)
-    {
+    auto it = _integrals.find({type, id, kernel_idx});
+    if (it == _integrals.end())
       throw std::runtime_error("No kernel for requested domain index.");
-    }
-
-    return std::next(start, kernel_idx)->entities;
+    return it->second.entities;
   }
 
-  /// @brief Compute the list of entity indices in `mesh` for the ith
-  /// integral (kernel) of a given type (i.e. cell, exterior facet, or
-  /// interior facet).
+  /// @brief Argument domain indices.
+  ///
+  /// The indices in the integration domain mesh that the argument
+  /// function, e.g. a test or trial function, is defined on.
   ///
   /// @param type Integral type.
-  /// @param i Integral ID, i.e. the (sub)domain index.
-  /// @param kernel_idx Index of the kernel (we may have multiple
-  /// kernels for a given ID in mixed-topology meshes).
-  /// @param mesh The mesh the entities are numbered with respect to.
-  /// @return List of active entities in `mesh` for the given integral.
-  std::vector<std::int32_t> domain(IntegralType type, int i, int kernel_idx,
-                                   const mesh::Mesh<geometry_type>& mesh) const
+  /// @param rank Argument index, e.g. `0` for test function space, `1`
+  /// for trial function space.
+  /// @param id Integral identifier index.
+  /// @param kernel_idx Kernel index (cell type).
+  /// @return An array `domain`, such that:
+  /// - For cell integrals, `domain[i]` is the index of the ith cell `i`
+  /// in the integration domain mesh.
+  /// - For exterior/interior facet integrals, `domain[2*i]` is the
+  /// index of a cell in the integration domain mesh attached to the
+  /// facets `i`, and `domain[2*i + 1]` is the local index of the facet
+  /// relative to the cell.
+  std::span<const std::int32_t> domain_arg(IntegralType type, int rank, int id,
```

Comment:
I don't think the docs for this function are correct. The returned list contains indices in the meshes that the trial or test functions are defined over, not the integration mesh.